### PR TITLE
Implement the freemod selection overlay

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/11665
- [x] https://github.com/ppy/osu/pull/11662

### CODE REVIEW

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 1]
```diff
@@ -494,19 +496,28 @@ private void updateMods()
...
+
         private void modButtonPressed(Mod selectedMod)
         {
             if (selectedMod != null)
             {
-                if (State.Value == Visibility.Visible) sampleOn?.Play();
+                // Fixes buzzing when multiple mods are selected in the same frame.
+                sampleOnDelegate?.Cancel();
+                if (State.Value == Visibility.Visible)
+                    sampleOnDelegate = Scheduler.Add(() => sampleOn?.Play());
```
Comment (bdach) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs:509]: I guess it's not a huge deal but you could presumably use `Scheduler.AddOnce()` for this, which would eliminate the delegate cancellation?

**Review by bdach — COMMENTED** [review]
Don't see many huge issues with this part.

File: osu.Game/Graphics/UserInterface/OsuCheckbox.cs [hunk 2]
```diff
@@ -18,6 +18,11 @@ public class OsuCheckbox : Checkbox
         public Color4 UncheckedColor { get; set; } = Color4.White;
         public int FadeDuration { get; set; }
 
+        /// <summary>
+        /// Whether to play sounds when the state changes as a result of user interaction.
+        /// </summary>
+        protected virtual bool PlaySoundsOnUserChange => true;
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/OsuCheckbox.cs:25]: What was the reasoning for this?

Comment (smoogipoo) [reply osu.Game/Graphics/UserInterface/OsuCheckbox.cs:25]: Because it's already playing from the mods being selected.

Comment (peppy) [reply osu.Game/Graphics/UserInterface/OsuCheckbox.cs:25]: Does that mean it's also playing 30 times when you click select all?

Comment (smoogipoo) [reply osu.Game/Graphics/UserInterface/OsuCheckbox.cs:25]: Originally it was, but see the changes to ModSelectOverlay.

Comment (peppy) [reply osu.Game/Graphics/UserInterface/OsuCheckbox.cs:25]: I see, thanks.

**Comment by peppy** [comment]
@smoogipoo sooo... i made a few changes. please check over and make sure you agree with things structurally and feel-wise. Hopefully not too much to make this harder to review (almost at the point I should have done it in a separate PR but oh well).

**Comment by smoogipoo** [comment]
Clicking DT then clicking HT deselects DT but doesn't select HT.

File: osu.Game/Graphics/UserInterface/OsuCheckbox.cs [hunk 3]
```diff
@@ -52,26 +58,40 @@ public OsuCheckbox()
...
+            if (nubOnRight)
+            {
+                Nub.Anchor = Anchor.CentreRight;
+                Nub.Origin = Anchor.CentreRight;
+                Nub.Margin = new MarginPadding { Right = nub_padding };
+                labelText.Padding = new MarginPadding { Right = Nub.EXPANDED_SIZE + nub_padding * 2 };
+            }
+            else
+            {
+                Nub.Margin = new MarginPadding { Left = nub_padding };
```
Comment (smoogipoo) [inline osu.Game/Graphics/UserInterface/OsuCheckbox.cs:79]: Needs anchors/origins like above:

image

**Comment by peppy** [comment]
The DT->HT one is a bit nasty (doesn't repro on test scenes; is related to the recently added mod validation logic). Will look into it.

**Comment by peppy** [comment]
I've added back the immediate deselection flow for this specific use case. I don't think there's a better way to handle this immediately.

**Comment by smoogipoo** [comment]
Seems to work fine, lgtm :+1:

**Review by peppy — APPROVED** [review]

**Pull request merged**